backend/agents/agent.py

```diff
@@ -0,0 +1,38 @@
+from abc import ABC
+from typing import List, Type
+from utils import call_model_with_tools
+from .adapters import get_mistral_tool
+from .types import Action_and_args, Tool
+
+
+class Agent(ABC):
+    name: str
+    description: str
+    tools: List[Tool]
+    prompt: str
+
+    def __get_action(self, utterance: str) -> Action_and_args:
+        tools = map(get_mistral_tool, self.tools)
+        (function_name, function_params) = call_model_with_tools(self.prompt, utterance, tools)
+
+        tool = next((tool for tool in self.tools if tool.action.__name__ == function_name), None)
+
+        if tool is None:
+            raise ValueError(f"Tool {function_name} not found in agent {self.name}")

```

Comment:
One for the future probably, but if we fail to pick a tool we should look at doing a retry LLM call with a log of what went wrong to see if it can resolve it's issue

Reply:
agree, and agree yeah one for the future.  Should be nice to change in here though and will apply for all agents!